### ISSUE TITLE
Make MaxConnectionDuration of core.HttpProtocolOptions configurable

### ISSUE THREAD

**Issue by pims** [projectcontour/contour#2247]
**Please describe the problem you have**
[A clear, concise, description of the problem you are facing. What is the _problem_ that feature X would solve for you?]

Problem we're facing: imposing sane limits to upstream/downstream connections for applications running in a large cluster with a shared ingress deployment.

This is a follow up issue from https://github.com/projectcontour/contour/issues/2225

From the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/protocol.proto#core-httpprotocoloptions):

> max_connection_duration
(Duration) The maximum duration of a connection. The duration is defined as a period since a connection was established. If not set, there is no max duration. When max_connection_duration is reached the connection will be closed. Drain sequence will occur prior to closing the connection if if’s applicable. See drain_timeout. Note: not implemented for upstream connections.

We run mixed workloads (http, gRPC) and have found that forcing a clean disconnect helps with connection reliability.

This is working well in conjunction with a short [drain-timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-drain-timeout).

Our settings:
 - max_connection_duration: `300s`
 - drain_timeout: `30s`

Setting a default value will likely change the currently observed behavior (no max value) and thus recommend making this setting configurable.

Happy to submit a PR with a proof of concept on how to make this value configurable if it’d be helpful.

**Comment by youngnick** [comment]
Thanks for logging this issue @pims. Can I ask, is the thing you really want here "make sure that not too many connections are running to backend services", and this is the way you've found works best for you?

The reason we don't set a default max_connection_duration is that it is a hard time limit, not an idle timeout or similar, which means that if you had a large, slow upload that would take longer than 300s, then it would never be able to succeed.

The reason I'm asking is that maybe one of Envoy's idle timeouts would be a better option for the use case I mentioned.

Just to be clear, I'm not saying "we won't do this", I just want to understand what your use of this duration is trying to achieve.

**Comment by pims** [comment]
Thanks for following up on this issue @youngnick.

That’s right, in our fork of Contour, we have added support for `max_connection_duration`, configured through a `--max-connection-duration=300s` flag. This forces a clean disconnect for clients that have been connected longer than the configured `max_connection_duration`.

By making the value configurable through a flag, we have been able to force clients to cleanly disconnects if they failed to complete their request in a timely manner.
This is one of the many Envoy knobs that we _had to_ tune to provide a good QoS in a multi-tenant, mixed workload, busy cluster.

I’ve listed a few other configuration changes we’ve made in https://github.com/projectcontour/contour/issues/2225 and was hoping to continue the discussion on exposing a few more settings for Envoy.

**Comment by youngnick** [comment]
I guess the question I was really asking was "Why not an idle timeout of some sort?". Using the hard connection timeout seems like a very blunt instrument to me.

**Comment by pims** [comment]
Oh, we’ve implemented both. They're both useful in different situations.
To also be honest, any timeout that can be made configurable would be great, I just happened to pick one among all of them. The reason I brought up many different timeouts was to help inform a potential design decision on how to configure multiple timeouts, as opposed to looking at only making one configurable at a time.

I'm happy to open a new issue on making `idle-timeout` configurable if you believe it's a better starting point.

**Comment by youngnick** [comment]
I guess what I'd like is to know more about how they're useful in different situations. What are the situations, and how are the different things useful? Probably better to take it back to #2225 since we're talking more generally though.

**Comment by pims** [comment]
@youngnick `max_connection_duration` is useful to terminate connections that are busy / never idle.

Imagine the following client:
```go
conn, _ := net.Dial("tcp", ":"+os.Args[1])
req := "GET /foo HTTP/1.1\nHost: example.com\nConnection: Keep-Alive\n\n"

b := make([]byte, 1024)
for {
	_, err = conn.Write([]byte(req))
	if err != nil {
		log.Fatal("write err", err)
        }
	n, err := conn.Read(b)
	if err != nil {
		log.Fatal(err)
        }
	log.Println(string(b[0:n]))
	time.Sleep(1 * time.Second)
}
```

If we only configure:
```yaml
common_http_protocol_options:
  idle_timeout: 10s
```
clients connected to this Envoy instance will "never" disconnect because they are never idle. To help with balancing load and clients, having `max_connection_duration` configured, forces Envoy to close the connection.

Before reaching the `max_connection_duration`, Envoy responds with:
```
':status', '200'
'date', 'Fri, 17 Apr 2020 14:29:54 GMT'
'content-length', '6'
'content-type', 'text/plain; charset=utf-8'
'x-envoy-upstream-service-time', '0'
'server', 'envoy'
```

then once `max_connection_duration` is reached:
```
[2020-04-17 07:29:55.671][3991128][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:495] [C0] max connection duration reached
```

after the `drain_timeout` (default 5s), Envoy responds with:

```
':status', '200'
'date', 'Fri, 17 Apr 2020 14:29:56 GMT'
'content-length', '6'
'content-type', 'text/plain; charset=utf-8'
'x-envoy-upstream-service-time', '1'
'server', 'envoy'
'connection', 'close' 👈🏼
```

and the client gets an expected:
```
2020/04/17 07:29:58 read tcp 127.0.0.1:54510->127.0.0.1:10000: read: connection reset by peer
```
which they hopefully handle and reconnect.

Hopefully that explains why I think both are good candidates for tunable timeout configuration, as they solve different use cases.

Let me know if I can provide more info.

**Comment by skriss** [comment]
this was closed by #2670.
